File: java/app/src/main/java/com/huawei/locationsample6/location/fusedlocation/GeocoderActivity.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2021. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.huawei.locationsample6.location.fusedlocation;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;

import com.huawei.hmf.tasks.OnFailureListener;
import com.huawei.hmf.tasks.OnSuccessListener;
import com.huawei.hms.location.GeocoderService;
import com.huawei.hms.location.GetFromLocationNameRequest;
import com.huawei.hms.location.GetFromLocationRequest;
import com.huawei.hms.location.HWLocation;
import com.huawei.hms.location.LocationServices;
import com.huawei.location.lite.common.util.ExecutorUtil;
import com.huawei.location.lite.common.util.GsonUtil;
import com.huawei.locationsample6.NoDoubleClickUtils;
import com.huawei.locationsample6.R;
import com.huawei.logger.LocationLog;

import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Geocoder
 *
 * @since 2021-11-23
 */
public class GeocoderActivity extends LocationBaseActivity implements View.OnClickListener {
    private static final String TAG = "ReverseGeoCodeActivity";

    private EditText longitude;

    private EditText latitude;

    private EditText maxResults;

    private EditText locationName;

    private EditText revGeoLanguage;

    private EditText revGeoCountry;

    private EditText geoLanguage;

    private EditText geoCountry;

    private LinearLayout reverseGeoCodeLayout;

    private LinearLayout geoCodeLayout;

    private EditText lowerLeftLatitude;

    private EditText lowerLeftLongitude;

    private EditText upperRightLatitude;

    private EditText upperRightLongitude;

    private EditText geoResults;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_geocoder);
        findViewById(R.id.reverseGeoCode).setOnClickListener(this);
        latitude = findViewById(R.id.latitude_value);
        longitude = findViewById(R.id.longitude_value);
        maxResults = findViewById(R.id.resultCount);
        locationName = findViewById(R.id.name);
        revGeoLanguage = findViewById(R.id.reverse_geocode_language);
        revGeoCountry = findViewById(R.id.reverse_geocode_country);
        reverseGeoCodeLayout = findViewById(R.id.reverseGeoCode_layout);
        geoCodeLayout = findViewById(R.id.geocode_layout);
        lowerLeftLatitude = findViewById(R.id.lowerLeftLatitude_input);
        lowerLeftLongitude = findViewById(R.id.lowerLeftLongitude_input);
        upperRightLatitude = findViewById(R.id.upperRightLatitude_input);
        upperRightLongitude = findViewById(R.id.upperRightLongitude_input);
        geoResults = findViewById(R.id.geocode_results);
        geoLanguage = findViewById(R.id.geocode_language);
        geoCountry = findViewById(R.id.geocode_country);
        addLogFragment();
    }

    @Override
    public void onClick(View v) {
        Log.i(TAG, "onClick: ");
        if (NoDoubleClickUtils.isFastDoubleClick()) {
            LocationLog.i(TAG, "please retry later!!!");
            return;
        }
        String latStr = latitude.getText().toString();
        String longiStr = longitude.getText().toString();
        String maxStr = maxResults.getText().toString();
        if (!checkLatValid(latStr)) {
            LocationLog.e(TAG, "Latitude is illegal");
            return;
        }
        if (!checkLongiValid(longiStr)) {
            LocationLog.e(TAG, "Longitude is illegal");
            return;
        }
        if (!isInt(maxStr)) {
            LocationLog.e(TAG, "maxResults is illegal");
            return;
        }
        final double lat = Double.parseDouble(latStr);
        final double longi = Double.parseDouble(longiStr);
        final int maxResult;
        try {
            maxResult = Integer.parseInt(maxStr);
        } catch (NumberFormatException e) {
            LocationLog.e(TAG, "maxResults is illegal");
            return;
        }
        final String language = revGeoLanguage.getText().toString();
        final String country = revGeoCountry.getText().toString();
        ExecutorUtil.getInstance().execute(new Runnable() {
            @Override
            public void run() {
                // Enter a proper region longitude and latitude. Otherwise, no geographic information is returned. If a
                // non-China region is used, transfer the longitude and latitude of the non-China region and ensure that
                // the longitude and latitude are correct.
                GetFromLocationRequest getFromLocationRequest = new GetFromLocationRequest(lat, longi, maxResult);
                Locale locale = new Locale(language, country);
                GeocoderService geocoderService = LocationServices.getGeocoderService(GeocoderActivity.this, locale);
                geocoderService.getFromLocation(getFromLocationRequest)
                    .addOnSuccessListener(new OnSuccessListener<List<HWLocation>>() {
                        @Override
                        public void onSuccess(List<HWLocation> hwLocation) {
                            printGeocoderResult(hwLocation);
                        }
                    })
                    .addOnFailureListener(new OnFailureListener() {
                        @Override
                        public void onFailure(Exception e) {
                            LocationLog.i(TAG, e.getMessage());
                        }
                    });
            }
        });
    }

    /**
     * Check whether a string is double using a regular expression.
     *
     * @param input String
     * @return Boolean value
     */
    private boolean isDouble(String input) {
        if (isInt(input)) {
            return true;
        }
        Matcher mer = Pattern.compile("^[+-]?[1-9]\\d*\\.\\d*|0\\.\\d*[1-9]\\d*$").matcher(input);
        return mer.matches();
    }

    public static boolean isInt(String input) {
        Matcher mer = Pattern.compile("^[+-]?[0-9]+$").matcher(input);
        return mer.matches();
    }

    private boolean checkLatValid(String latitude) {
        if (!isDouble(latitude)) {
            return false;
        }
        return true;
    }

    private boolean checkLongiValid(String longitude) {
        if (!isDouble(longitude)) {
            return false;
        }
        return true;
    }

    public void geoCoder(View view) {
        if (NoDoubleClickUtils.isFastDoubleClick()) {
            LocationLog.i(TAG, "please retry later!!!");
            return;
        }

        if (!isInt(geoResults.getText().toString())) {
            LocationLog.e(TAG, "maxResults is illegal");
            return;
        }
        final String addressName = locationName.getText().toString();
        final int results;
        try {
            results = Integer.parseInt(geoResults.getText().toString());
        } catch (Exception e) {
            LocationLog.e(TAG, "maxResults is illegal");
            return;
        }
        final String language = geoLanguage.getText().toString();
        final String country = geoCountry.getText().toString();
        ExecutorUtil.getInstance().execute(new Runnable() {
            @Override
            public void run() {
                GetFromLocationNameRequest getFromLocationNameRequest =
                    new GetFromLocationNameRequest(addressName, results);
                if (checkLatValid(lowerLeftLatitude.getText().toString())
                    && checkLongiValid(lowerLeftLongitude.getText().toString())
                    && checkLatValid(upperRightLatitude.getText().toString())
                    && checkLongiValid(upperRightLongitude.getText().toString())) {
                    double lowerLeftlat = Double.parseDouble(lowerLeftLatitude.getText().toString());
                    double lowerLeftlng = Double.parseDouble(lowerLeftLongitude.getText().toString());
                    double upperrightlat = Double.parseDouble(upperRightLatitude.getText().toString());
                    double upperrightlng = Double.parseDouble(upperRightLongitude.getText().toString());
                    getFromLocationNameRequest.setLowerLeftLatitude(lowerLeftlat);
                    getFromLocationNameRequest.setLowerLeftLongitude(lowerLeftlng);
                    getFromLocationNameRequest.setUpperRightLatitude(upperrightlat);
                    getFromLocationNameRequest.setUpperRightLongitude(upperrightlng);
                }
                Locale locale = new Locale(language, country);
                GeocoderService geocoderService = LocationServices.getGeocoderService(GeocoderActivity.this, locale);
                // Enter the correct location information. Otherwise, the geographic information cannot be parsed. For a
                // non-China region, transfer the location information of the non-China region.
                geocoderService.getFromLocationName(getFromLocationNameRequest)
                    .addOnSuccessListener(new OnSuccessListener<List<HWLocation>>() {
                        @Override
                        public void onSuccess(List<HWLocation> hwLocations) {
                            printGeocoderResult(hwLocations);
                        }
                    })
                    .addOnFailureListener(new OnFailureListener() {
                        @Override
                        public void onFailure(Exception ex) {
                            LocationLog.i(TAG, ex.getMessage());
                        }
                    });
            }
        });
    }

    private void printGeocoderResult(List<HWLocation> geocoderResult) {
        if (geocoderResult == null || geocoderResult.isEmpty()) {
            LocationLog.i(TAG, GsonUtil.getInstance().toJson(geocoderResult));
        } else {
            for (HWLocation hwLocation : geocoderResult) {
                StringBuilder builder = new StringBuilder();
                if (hwLocation != null) {

                    builder.append("Location:{latitude=")
                        .append(hwLocation.getLatitude())
                        .append(",longitude=")
                        .append(hwLocation.getLongitude())
                        .append(",countryName=")
                        .append(hwLocation.getCountryName())
                        .append(",countryCode=")
                        .append(hwLocation.getCountryCode())
                        .append(",state=")
                        .append(hwLocation.getState())
                        .append(",city=")
                        .append(hwLocation.getCity())
                        .append(",county=")
                        .append(hwLocation.getCounty())
                        .append(",street=")
                        .append(hwLocation.getStreet())
                        .append(",featureName=")
                        .append(hwLocation.getFeatureName())
                        .append(",postalCode=")
                        .append(hwLocation.getPostalCode())
                        .append(",phone=")
                        .append(hwLocation.getPhone())
                        .append(",url=")
                        .append(hwLocation.getUrl())
                        .append(",extraInfo=")
                        .append(GsonUtil.getInstance().toJson(hwLocation.getExtraInfo()))
                        .append("}" + System.lineSeparator());
                } else {
                    builder.append("Location:{}");
                }
                LocationLog.i(TAG, builder.toString());
            }
        }
    }

    public void switchMode(View view) {
        if (reverseGeoCodeLayout.isShown()) {
            reverseGeoCodeLayout.setVisibility(View.GONE);
            geoCodeLayout.setVisibility(View.VISIBLE);

        } else {
            geoCodeLayout.setVisibility(View.GONE);
            reverseGeoCodeLayout.setVisibility(View.VISIBLE);
        }
    }

}

File: hms-location-demo-android-studio/kotlin/app/src/main/java/com/hms/locationsample6/geofence/GeocoderActivity.kt
/**
 * Copyright 2020 Huawei Technologies co, Ltd All
 * Rights reserved
 * Licenced under the Apache License,Version 2.0(the "License");
 * you may not use this file except in compliance with license
 * you may obtain a copy of the license at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by application law or agreed to in writing software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permission and
 * limitations under the License
 */

package com.hms.locationsample6.geofence

import android.os.Bundle
import android.view.View
import android.widget.EditText
import android.widget.LinearLayout
import com.hms.locationsample6.R
import com.hms.locationsample6.activity.BaseActivity
import com.hms.locationsample6.logger.LocationLog
import com.hms.locationsample6.utils.NoDoubleClickUtils
import com.huawei.hmf.tasks.OnSuccessListener
import com.huawei.hms.location.GetFromLocationNameRequest
import com.huawei.hms.location.GetFromLocationRequest
import com.huawei.hms.location.HWLocation
import com.huawei.hms.location.LocationServices
import com.huawei.location.lite.common.util.ExecutorUtil
import com.huawei.location.lite.common.util.GsonUtil
import kotlinx.android.synthetic.main.activity_geocoder.*
import java.util.*
import java.util.regex.Pattern

class GeocoderActivity : BaseActivity(), View.OnClickListener {
    private val TAG = "ReverseGeoCodeActivity"

    private var longitude: EditText? = null

    private var latitude: EditText? = null

    private var maxResults: EditText? = null

    private var locationName: EditText? = null

    private var revGeoLanguage: EditText? = null

    private var revGeoCountry: EditText? = null

    private var geoLanguage: EditText? = null

    private var geoCountry: EditText? = null

    private var reverseGeoCodeLayout: LinearLayout? = null

    private var geoCodeLayout: LinearLayout? = null

    private var lowerLeftLatitude: EditText? = null

    private var lowerLeftLongitude: EditText? = null

    private var upperRightLatitude: EditText? = null

    private var upperRightLongitude: EditText? = null

    private var geoResults: EditText? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_geocoder)
        btn_reverseGeoCode?.setOnClickListener(this)
        btn_switch_mode?.setOnClickListener(this)
        btn_geocode?.setOnClickListener(this)
        latitude = findViewById(R.id.latitude_value)
        longitude = findViewById(R.id.longitude_value)
        maxResults = findViewById(R.id.resultCount)
        locationName = findViewById(R.id.name)
        revGeoLanguage = findViewById(R.id.reverse_geocode_language)
        revGeoCountry = findViewById(R.id.reverse_geocode_country)
        reverseGeoCodeLayout = findViewById(R.id.reverseGeoCode_layout)
        geoCodeLayout = findViewById(R.id.geocode_layout)
        lowerLeftLatitude = findViewById(R.id.lowerLeftLatitude_input)
        lowerLeftLongitude = findViewById(R.id.lowerLeftLongitude_input)
        upperRightLatitude = findViewById(R.id.upperRightLatitude_input)
        upperRightLongitude = findViewById(R.id.upperRightLongitude_input)
        geoResults = findViewById(R.id.geocode_results)
        geoLanguage = findViewById(R.id.geocode_language)
        geoCountry = findViewById(R.id.geocode_country)
        addLogFragment()
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.btn_reverseGeoCode -> reverseGeoCode()
            R.id.btn_switch_mode -> switchMode()
            R.id.btn_geocode -> geoCoder()

        }
    }

    private fun reverseGeoCode() {
        if (NoDoubleClickUtils.isFastDoubleClick()) {
            LocationLog.i(TAG, "please retry later!!!")
            return
        }
        val latStr = latitude!!.text.toString()
        val longiStr = longitude!!.text.toString()
        val maxStr = maxResults!!.text.toString()
        if (!checkLatValid(latStr)) {
            LocationLog.e(TAG, "Latitude is illegal")
            return
        }
        if (!checkLongiValid(longiStr)) {
            LocationLog.e(TAG, "Longitude is illegal")
            return
        }
        if (!isInt(maxStr)) {
            LocationLog.e(TAG, "maxResults is illegal")
            return
        }
        val lat = latStr.toDouble()
        val longi = longiStr.toDouble()
        val maxResult: Int
        maxResult = try {
            maxStr.toInt()
        } catch (e: NumberFormatException) {
            LocationLog.e(TAG, "maxResults is illegal")
            return
        }
        val language = revGeoLanguage!!.text.toString()
        val country = revGeoCountry!!.text.toString()
        ExecutorUtil.getInstance().execute {
            val getFromLocationRequest =
                GetFromLocationRequest(lat, longi, maxResult)
            val locale = Locale(language, country)
            val geocoderService =
                LocationServices.getGeocoderService(this@GeocoderActivity, locale)
            geocoderService.getFromLocation(getFromLocationRequest)
                .addOnSuccessListener { hwLocation -> printGeocoderResult(hwLocation) }
                .addOnFailureListener { e ->
                    LocationLog.i(
                        TAG,
                        "getFromLocation fail:${e.message}"
                    )
                }
        }
    }

    private fun switchMode() {
        if (reverseGeoCodeLayout!!.isShown) {
            reverseGeoCodeLayout!!.visibility = View.GONE
            geoCodeLayout!!.visibility = View.VISIBLE
        } else {
            geoCodeLayout!!.visibility = View.GONE
            reverseGeoCodeLayout!!.visibility = View.VISIBLE
        }
    }

    private fun geoCoder() {
        if (NoDoubleClickUtils.isFastDoubleClick()) {
            LocationLog.i(TAG, "please retry later!!!")
            return
        }
        if (!isInt(geoResults!!.text.toString())) {
            LocationLog.e(TAG, "maxResults is illegal")
            return
        }
        val addressName = locationName!!.text.toString()
        val results: Int
        results = try {
            geoResults!!.text.toString().toInt()
        } catch (e: Exception) {
            LocationLog.e(TAG, "maxResults is illegal")
            return
        }
        val language = geoLanguage!!.text.toString()
        val country = geoCountry!!.text.toString()
        ExecutorUtil.getInstance().execute {
            val getFromLocationNameRequest =
                GetFromLocationNameRequest(addressName, results)
            if (checkLatValid(lowerLeftLatitude!!.text.toString())
                && checkLongiValid(lowerLeftLongitude!!.text.toString())
                && checkLatValid(upperRightLatitude!!.text.toString())
                && checkLongiValid(upperRightLongitude!!.text.toString())
            ) {
                val lowerLeftLat =
                    lowerLeftLatitude!!.text.toString().toDouble()
                val lowerLeftLng =
                    lowerLeftLongitude!!.text.toString().toDouble()
                val upperRightLat =
                    upperRightLatitude!!.text.toString().toDouble()
                val upperRightLng =
                    upperRightLongitude!!.text.toString().toDouble()
                getFromLocationNameRequest.lowerLeftLatitude = lowerLeftLat
                getFromLocationNameRequest.lowerLeftLongitude = lowerLeftLng
                getFromLocationNameRequest.upperRightLatitude = upperRightLat
                getFromLocationNameRequest.upperRightLongitude = upperRightLng
            }
            val locale = Locale(language, country)
            val geocoderService =
                LocationServices.getGeocoderService(this@GeocoderActivity, locale)
            geocoderService.getFromLocationName(getFromLocationNameRequest)
                .addOnSuccessListener(OnSuccessListener<List<HWLocation>?> { hwLocations ->
                    printGeocoderResult(
                        hwLocations
                    )
                })
                .addOnFailureListener { ex ->
                    LocationLog.i(
                        TAG,
                        "getFromLocationName fail:${ex.message}"
                    )
                }
        }
    }

    /**
     * Check whether a string is double using a regular expression.
     *
     * @param input String
     * @return Boolean value
     */
    private fun isDouble(input: String): Boolean {
        if (isInt(input)) {
            return true
        }
        val mer =
            Pattern.compile("^[+-]?[1-9]\\d*\\.\\d*|0\\.\\d*[1-9]\\d*$")
                .matcher(input)
        return mer.matches()
    }

    fun isInt(input: String?): Boolean {
        val mer =
            Pattern.compile("^[+-]?[0-9]+$").matcher(input)
        return mer.matches()
    }

    private fun checkLatValid(latitude: String): Boolean {
        return isDouble(latitude)
    }

    private fun checkLongiValid(longitude: String): Boolean {
        return isDouble(longitude)
    }


    private fun printGeocoderResult(geocoderResult: List<HWLocation>?) {
        if (geocoderResult == null || geocoderResult.isEmpty()) {
            LocationLog.i(TAG, GsonUtil.getInstance().toJson(geocoderResult))
        } else {
            for (hwLocation in geocoderResult) {
                val builder = StringBuilder()
                if (hwLocation != null) {
                    builder.append("Location:{latitude=")
                        .append(hwLocation.latitude)
                        .append(",longitude=")
                        .append(hwLocation.longitude)
                        .append(",countryName=")
                        .append(hwLocation.countryName)
                        .append(",countryCode=")
                        .append(hwLocation.countryCode)
                        .append(",state=")
                        .append(hwLocation.state)
                        .append(",city=")
                        .append(hwLocation.city)
                        .append(",county=")
                        .append(hwLocation.county)
                        .append(",street=")
                        .append(hwLocation.street)
                        .append(",featureName=")
                        .append(hwLocation.featureName)
                        .append(",postalCode=")
                        .append(hwLocation.postalCode)
                        .append(",phone=")
                        .append(hwLocation.phone)
                        .append(",url=")
                        .append(hwLocation.url)
                        .append(",extraInfo=")
                        .append(GsonUtil.getInstance().toJson(hwLocation.extraInfo))
                        .append("}" + System.lineSeparator())
                } else {
                    builder.append("Location:{}")
                }
                LocationLog.i(TAG, builder.toString())
            }
        }
    }
}